hash/hash_comparator.go

```diff
@@ -106,11 +109,35 @@ func ComparePbkdf2(_ context.Context, password []byte, hash []byte) error {
 	return errors.WithStack(ErrMismatchedHashAndPassword)
 }
 
+func CompareScrypt(_ context.Context, password []byte, hash []byte) error {
+	// Extract the parameters, salt and derived key from the encoded password
+	// hash.
+	p, salt, hash, err := decodeScryptHash(string(hash))
+	if err != nil {
+		return err
+	}
+
+	// Derive the key from the other password using the same parameters.
+	otherHash, err := scrypt.Key(password, salt, int(p.Cost), int(p.Block), int(p.Parrellization), int(p.KeyLength))
+	if err != nil {
+		return err
+	}
+
+	// Check that the contents of the hashed passwords are identical. Note
+	// that we are using the subtle.ConstantTimeCompare() function for this
+	// to help prevent timing attacks.
+	if subtle.ConstantTimeCompare(hash, otherHash) == 1 {
+		return nil
+	}
+	return errors.WithStack(ErrMismatchedHashAndPassword)
```

Comment:
Here we do need a test to check what happens if a password is supplied that does not match its hash.

Reply:
Added the test [here](https://github.com/ory/kratos/pull/2689/files#diff-275881c2b9629b0999bee794700468eea536ffa1e5b2078beb854816435b4c30R244), same for the one before, the codecov warning is still there but the test are ok

Reply:
Done [here](https://github.com/ory/kratos/pull/2689/files#diff-275881c2b9629b0999bee794700468eea536ffa1e5b2078beb854816435b4c30R243)